### ISSUE TITLE
Node broken when install-cni fails

### ISSUE THREAD

**Issue by sathieu** [istio/istio#36821]
### Bug Description

<summary>

The `install-cni` pod failed to install the `cni-node` binary because of "too many open files":

<details>

```
$ kubectl logs -n istio-system istio-cni-node-l248n
2022-01-13T08:59:44.319301Z     info    CNI install configuration:
CNINetDir: /etc/cni/net.d
MountedCNINetDir: /host/etc/cni/net.d
CNIConfName:
ChainedCNIPlugin: true
CNINetworkConfigFile:
CNINetworkConfig: {
  "cniVersion": "0.3.1",
  "name": "istio-cni",
  "type": "istio-cni",
  "log_level": "info",
  "log_uds_address": "__LOG_UDS_ADDRESS__",
  "kubernetes": {
      "kubeconfig": "__KUBECONFIG_FILEPATH__",
      "cni_bin_dir": "/opt/cni/bin",
      "exclude_namespaces": [ "istio-system", "kube-system" ]
  }
}
CNIEnableReinstall: true
LogLevel: warn
KubeconfigFilename: ZZZ-istio-cni-kubeconfig
KubeconfigMode: 0600
KubeCAFile:
SkipTLSVerify: false
K8sServiceProtocol:
K8sServiceHost: 10.233.0.1
K8sServicePort: 443
K8sNodeName: istio-cni-node-l248n
UpdateCNIBinaries: true
SkipCNIBinaries: [[]]
MonitoringPort: 15014
LogUDSAddress: /var/run/istio-cni/log.sock

2022-01-13T08:59:44.319371Z     info    CNI race repair configuration:
Enabled: true
RunAsDaemon: true
NodeName: --REDACTED--
LabelKey: cni.istio.io/uninitialized
LabelValue: true
DeletePods: true
LabelPods: true
SidecarAnnotation: sidecar.istio.io/status
InitContainerName: istio-validation
InitTerminationMsg:
InitExitCode: 126
LabelSelectors:
FieldSelectors:

2022-01-13T08:59:44.319368Z     info    ControlZ available at 127.0.0.1:9876
2022-01-13T08:59:44.321040Z     info    Start a UDS server for CNI plugin logs
2022-01-13T08:59:44.321349Z     info    repair  Start CNI race condition repair.
2022-01-13T08:59:45.259203Z     info    install Copied istio-cni to /host/opt/cni/bin.
2022-01-13T08:59:45.259877Z     info    install Directory /host/secondary-bin-dir is not writable, skipping.
2022-01-13T08:59:45.260213Z     info    install write kubeconfig file /host/etc/cni/net.d/ZZZ-istio-cni-kubeconfig with:
# Kubeconfig file for Istio CNI plugin.
apiVersion: v1
kind: Config
clusters:
- name: local
  cluster:
    server: https://[10.233.0.1]:443
    certificate-authority-data: <CA cert from /var/run/secrets/kubernetes.io/serviceaccount/ca.crt>
users:
- name: istio-cni
  user:
    token: "<redacted>"
contexts:
- name: istio-cni-context
  context:
    cluster: local
    user: istio-cni
current-context: istio-cni-context

2022-01-13T08:59:45.260593Z     info    install Using CNI config template from CNI_NETWORK_CONFIG environment variable.
2022-01-13T08:59:45.260636Z     info    install CNI config: {
  "cniVersion": "0.3.1",
  "name": "istio-cni",
  "type": "istio-cni",
  "log_level": "info",
  "log_uds_address": "/var/run/istio-cni/log.sock",
  "kubernetes": {
      "kubeconfig": "/etc/cni/net.d/ZZZ-istio-cni-kubeconfig",
      "cni_bin_dir": "/opt/cni/bin",
      "exclude_namespaces": [ "istio-system", "kube-system" ]
  }
}
2022-01-13T08:59:45.260729Z     error   Installer exits with too many open files
2022-01-13T08:59:45.260740Z     info    install Cleaning up.
2022-01-13T08:59:45.260758Z     info    install Removing Istio CNI kubeconfig file: /host/etc/cni/net.d/ZZZ-istio-cni-kubeconfig
2022-01-13T08:59:45.260864Z     info    install Removing binary: /host/opt/cni/bin/istio-cni
Error: too many open files
Usage:
  install-cni [flags]
  install-cni [command]

Available Commands:
  completion  generate the autocompletion script for the specified shell
  help        Help about any command
  version     Prints out build version information

Flags:
      --chained-cni-plugin                                 Whether to install CNI plugin as a chained or standalone (default true)
      --cni-conf-name string                               Name of the CNI configuration file
      --cni-enable-reinstall                               Whether to reinstall CNI configuration and binary files (default true)
      --cni-net-dir string                                 Directory on the host where CNI network plugins are installed (default "/etc/cni/net.d")
      --cni-network-config string                          CNI configuration template as a string
      --cni-network-config-file string                     CNI config template as a file
      --ctrlz_address string                               The IP Address to listen on for the ControlZ introspection facility. Use '*' to indicate all addresses. (default "localhost")
      --ctrlz_port uint16                                  The IP port to use for the ControlZ introspection facility (default 9876)
  -h, --help                                               help for install-cni
      --kube-ca-file string                                CA file for kubeconfig. Defaults to the same as install-cni pod
      --kubecfg-file-name string                           Name of the kubeconfig file which CNI plugin will use when interacting with API server (default "ZZZ-istio-cni-kubeconfig")
      --kubeconfig-mode int                                File mode of the kubeconfig file (default 384)
      --log-level string                                   Fallback value for log level in CNI config file, if not specified in helm template (default "warn")
      --log-uds-address string                             The UDS server address which CNI plugin will copy log ouptut to (default "/var/run/istio-cni/log.sock")
      --log_as_json                                        Whether to format output as JSON or in plain console-friendly format
      --log_caller string                                  Comma-separated list of scopes for which to include caller information, scopes can be any of [all, cni, default, install, klog, repair]
      --log_output_level string                            Comma-separated minimum per-scope logging level of messages to output, in the form of <scope>:<level>,<scope>:<level>,... where scope can be one of [all, cni, default, install, klog, repair] and level can be one of [debug, info, warn, error, fatal, none] (default "default:info")
      --log_rotate string                                  The path for the optional rotating log file
      --log_rotate_max_age int                             The maximum age in days of a log file beyond which the file is rotated (0 indicates no limit) (default 30)
      --log_rotate_max_backups int                         The maximum number of log file backups to keep before older files are deleted (0 indicates no limit) (default 1000)
      --log_rotate_max_size int                            The maximum size in megabytes of a log file beyond which the file is rotated (default 104857600)
      --log_stacktrace_level string                        Comma-separated minimum per-scope logging level at which stack traces are captured, in the form of <scope>:<level>,<scope:level>,... where scope can be one of [all, cni, default, install, klog, repair] and level can be one of [debug, info, warn, error, fatal, none] (default "default:none")
      --log_target stringArray                             The set of paths where to output the log. This can be any path as well as the special values stdout and stderr (default [stdout])
      --monitoring-port int                                HTTP port to serve prometheus metrics (default 15014)
      --mounted-cni-net-dir string                         Directory on the container where CNI networks are installed (default "/host/etc/cni/net.d")
      --repair-broken-pod-label-key string                 The key portion of the label which will be set by the ace repair if label pods is true (default "cni.istio.io/uninitialized")
      --repair-broken-pod-label-value string               The value portion of the label which will be set by the race repair if label pods is true (default "true")
      --repair-delete-pods                                 Controller will delete pods when detecting pod broken by race condition
      --repair-enabled                                     Whether to enable race condition repair or not (default true)
      --repair-field-selectors string                      A set of field selectors in label=value format that will be added to the pod list filters
      --repair-init-container-exit-code int                Expected exit code for the init container when crash-looping because of CNI misconfiguration (default 126)
      --repair-init-container-name string                  The name of the istio init container (will crash-loop if CNI is not configured for the pod) (default "istio-validation")
      --repair-init-container-termination-message string   The expected termination message for the init container when crash-looping because of CNI misconfiguration
      --repair-label-pods                                  Controller will label pods when detecting pod broken by race condition
      --repair-label-selectors string                      A set of label selectors in label=value format that will be added to the pod list filters
      --repair-node-name string                            The name of the managed node (will manage all nodes if unset)
      --repair-run-as-daemon                               Controller will run in a loop
      --repair-sidecar-annotation string                   An annotation key that indicates this pod contains an istio sidecar. All pods without this annotation will be ignored.The value of the annotation is ignored. (default "sidecar.istio.io/status")
      --skip-cni-binaries istio-cni                        Binaries that should not be installed. Currently Istio only installs one binary istio-cni
      --skip-tls-verify                                    Whether to use insecure TLS in kubeconfig file
      --update-cni-binaries                                Whether to refresh existing binaries when installing CNI (default true)

Use "install-cni [command] --help" for more information about a command.

```
</details>
</summary>

I tried to delete this pod, and it is blocked because of the missing binary:

```
$ kubectl describe -n istio-system istio-cni-node-l248n
[...]
  Warning  FailedKillPod  2m34s (x179 over 41m)  kubelet  error killing pod: failed to "KillPodSandbox" for "f504d3b1-0eff-48b1-9100-29f2946c36b7" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to destroy network for sandbox \"b128787d2dbd72d12c0efa0fe2427ecd16e5311680977126c17cd5c254d0480c\": failed to find plugin \"istio-cni\" in path [/opt/cni/bin]"
```

NB: All pods in the Terminating state on this node have the same error.

Question: How can I repair this node (should I force-delete the pod)?

Bug: a failure in the `install-cni` pod shouldn't break the node.

### Version

```prose
$ istioctl version
client version: 1.12.1
control plane version: 1.12.1
data plane version: 1.12.1 (56 proxies)
$ kubectl version --short
Client Version: v1.21.5
Server Version: v1.22.5
```

**Comment by sathieu** [comment]
NB: I'm unable to get ulimits on this broken pod, but I have the following on another one:

```
$ kubectl exec -it -n istio-system istio-cni-node-phtzr -- /bin/sh -c 'ulimit -a'
time(seconds)        unlimited
file(blocks)         unlimited
data(kbytes)         unlimited
stack(kbytes)        8192
coredump(blocks)     unlimited
memory(kbytes)       unlimited
locked memory(kbytes) 64
process              unlimited
nofiles              1048576
vmemory(kbytes)      unlimited
locks                unlimited
rtprio               0
```

**Comment by howardjohn** [comment]
> Bug: a failure in the install-cni pod shouldn't break the node.

Actually not sure I agree with this. If istio-cni is down, but pods are scheduled, they will not have iptables redirection setup which will bypass all security policies, etc.

> ulimits

CNI is coping files onto a hostpath mount btw, so it might be required to look at the node as well rather than just the pod.

**Comment by bianpengyuan** [comment]
The pod removal error is concerning. Looks like install-cni writes the cni config file, but did not clean it up if file watch errors out.

**Comment by bianpengyuan** [comment]
@sathieu Could you check what is the cni config looks like at the problematic node? It should be under `/etc/cni/net.d` if you did't set `values.cni.cniConfDir`. Also I guess the istio cni pod is in crashlooping state at that node? Is there any chance you can query the history log from the crash pod and see if there was any log line `CNI config file xxxx exists. Proceeding`? If so that would give more confidence that #36831 would be the right fix for the pod deletion issue.

**Comment by sathieu** [comment]
@bianpengyuan The istio CNI config and bin are not here, on the problematic node:

```console
$ sudo LANG=C ls -l /etc/cni/net.d /opt/cni/bin/istio-cni
ls: cannot access '/opt/cni/bin/istio-cni': No such file or directory
/etc/cni/net.d:
total 12
-rw-r--r-- 1 root root 1115 Jan 12 17:40 10-calico.conflist
-rw------- 1 root root 2854 Jan 17 12:03 calico-kubeconfig
-rw-r--r-- 1 root root  715 Sep 21 11:00 calico.conflist.template
```

Those are present on a working node:

```console
$ sudo ls -l /etc/cni/net.d /opt/cni/bin/istio-cni
-rwxr-xr-x 1 root root 68374528 janv. 17 12:02 /opt/cni/bin/istio-cni

/etc/cni/net.d:
total 16
-rw-r--r-- 1 root root 1115 janv. 17 12:02 10-calico.conflist
-rw-r--r-- 1 root root  715 sept. 21 11:00 calico.conflist.template
-rw------- 1 root root 2854 janv. 17 12:02 calico-kubeconfig
-rw------- 1 root root 2843 janv. 17 12:02 ZZZ-istio-cni-kubeconfig

```

**Comment by sathieu** [comment]
@bianpengyuan The istio-cni-node-* log is in the [issue description](https://github.com/istio/istio/issues/36821#issue-1101450920), you need to click on "Details" to expand the log.

**Comment by sathieu** [comment]
I recovered my node by copying `/opt/cni/bin/istio-cni` from another node. This allowed the `istio-cni-node-l248n` pod to terminate. A new pod was created (from the DaemonSet) without failures and `/etc/cni/net.d/ZZZ-istio-cni-kubeconfig` is back.

**Comment by sathieu** [comment]
Closing bug report. It looks like `istio-cni-node-` is not robust against `too many open files` errors.

At least #36831 will improve the message, and maybe avoid the failure.

Feel free to re-open if something is missing.

**Comment by bianpengyuan** [comment]
> @bianpengyuan The istio-cni-node-* log is in the [issue description](https://github.com/istio/istio/issues/36821#issue-1101450920), you need to click on "Details" to expand the log.

@sathieu Thanks! I guess the log was from the very recent crashes when you reported the bug? I was actually looking for logs from the pod that originally caused this issue (i.e. the pod which cleans up the cni binary but does not clean up the cni configuration). Likely it should be the very first run of the cni pod on this node.

**Comment by sathieu** [comment]
@bianpengyuan Unfortuantely, I've completely removed this test cluster. No way to get the logs back ...

**Comment by sathieu** [comment]
I'm reopening, because this bug is still present with install-cni 1.13.1 (the pod is Terminating because I deleted it while is was crashloopbackoff):


```shell
$ kubectl describe po -n istio-system istio-cni-node-q7d24
Name:                      istio-cni-node-q7d24
Namespace:                 istio-system
Priority:                  2000001000
Priority Class Name:       system-node-critical
Node:                      sqkubworkelst06/10.146.61.106
Start Time:                Tue, 08 Mar 2022 11:26:36 +0100
Labels:                    controller-revision-hash=5774898dd
                           k8s-app=istio-cni-node
                           pod-template-generation=3
                           sidecar.istio.io/inject=false
Annotations:               cni.projectcalico.org/containerID: 720259ea5a0427eee9bad6734ac0936308c8e2caf5a91be9339fcd1edfb0f3d8
                           cni.projectcalico.org/podIP: 10.233.126.67/32
                           cni.projectcalico.org/podIPs: 10.233.126.67/32
                           kubernetes.io/psp: istio-cni-node
                           prometheus.io/path: /metrics
                           prometheus.io/port: 15014
                           prometheus.io/scrape: true
                           scheduler.alpha.kubernetes.io/critical-pod:
                           sidecar.istio.io/inject: false
Status:                    Terminating (lasts 28m)
Termination Grace Period:  5s
IP:                        10.233.126.67
IPs:
  IP:           10.233.126.67
Controlled By:  DaemonSet/istio-cni-node
Containers:
  install-cni:
    Container ID:  containerd://eb5ca1f0c5fa887861c20a4b3dbb0bcc99d4ca2c5e1420a5c83b91df4def3296
    Image:         docker.io/istio/install-cni:1.13.1
    Image ID:      docker.io/istio/install-cni@sha256:e0515a519ebc567b267c2c3bb1e58283db2a4a6804c4ab74ea366f80dd4fbedb
    Port:          <none>
    Host Port:     <none>
    Command:
      install-cni
    State:          Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Wed, 09 Mar 2022 09:24:36 +0100
      Finished:     Wed, 09 Mar 2022 09:24:36 +0100
    Ready:          False
    Restart Count:  100
    Requests:
      cpu:      100m
      memory:   100Mi
    Liveness:   http-get http://:8000/healthz delay=5s timeout=1s period=10s #success=1 #failure=3
    Readiness:  http-get http://:8000/readyz delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:
      CNI_NETWORK_CONFIG:             <set to the key 'cni_network_config' of config map 'istio-cni-config'>  Optional: false
      CNI_NET_DIR:                    /etc/cni/net.d
      CHAINED_CNI_PLUGIN:             true
      REPAIR_ENABLED:                 true
      REPAIR_NODE_NAME:                (v1:spec.nodeName)
      REPAIR_LABEL_PODS:              true
      REPAIR_DELETE_PODS:             true
      REPAIR_RUN_AS_DAEMON:           true
      REPAIR_SIDECAR_ANNOTATION:      sidecar.istio.io/status
      REPAIR_INIT_CONTAINER_NAME:     istio-validation
      REPAIR_BROKEN_POD_LABEL_KEY:    cni.istio.io/uninitialized
      REPAIR_BROKEN_POD_LABEL_VALUE:  true
    Mounts:
      /host/etc/cni/net.d from cni-net-dir (rw)
      /host/opt/cni/bin from cni-bin-dir (rw)
      /var/run/istio-cni from cni-log-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-zfwtg (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  cni-bin-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /opt/cni/bin
    HostPathType:
  cni-net-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/cni/net.d
    HostPathType:
  cni-log-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/istio-cni
    HostPathType:
  kube-api-access-zfwtg:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              kubernetes.io/os=linux
Tolerations:                 :NoSchedule op=Exists
                             :NoExecute op=Exists
                             CriticalAddonsOnly op=Exists
                             node.kubernetes.io/disk-pressure:NoSchedule op=Exists
                             node.kubernetes.io/memory-pressure:NoSchedule op=Exists
                             node.kubernetes.io/not-ready:NoExecute op=Exists
                             node.kubernetes.io/pid-pressure:NoSchedule op=Exists
                             node.kubernetes.io/unreachable:NoExecute op=Exists
                             node.kubernetes.io/unschedulable:NoSchedule op=Exists
Events:
  Type     Reason         Age                  From     Message
  ----     ------         ----                 ----     -------
  Warning  BackOff        30m (x2421 over 8h)  kubelet  Back-off restarting failed container
  Warning  FailedKillPod  9s (x136 over 29m)   kubelet  error killing pod: failed to "KillPodSandbox" for "4c21b026-fae3-4597-9f0c-f157857f21d6" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to destroy network for sandbox \"720259ea5a0427eee9bad6734ac0936308c8e2caf5a91be9339fcd1edfb0f3d8\": failed to find plugin \"istio-cni\" in path [/opt/cni/bin]"

$ kubectl logs -n istio-system istio-cni-node-q7d24
2022-03-09T08:24:36.693521Z     info    ControlZ available at 127.0.0.1:9876
2022-03-09T08:24:36.694119Z     info    CNI install configuration:
CNINetDir: /etc/cni/net.d
MountedCNINetDir: /host/etc/cni/net.d
CNIConfName:
ChainedCNIPlugin: true
CNINetworkConfigFile:
CNINetworkConfig: {
  "cniVersion": "0.3.1",
  "name": "istio-cni",
  "type": "istio-cni",
  "log_level": "info",
  "log_uds_address": "__LOG_UDS_ADDRESS__",
  "kubernetes": {
      "kubeconfig": "__KUBECONFIG_FILEPATH__",
      "cni_bin_dir": "/opt/cni/bin",
      "exclude_namespaces": [ "istio-system", "kube-system" ]
  }
}
CNIEnableReinstall: true
LogLevel: warn
KubeconfigFilename: ZZZ-istio-cni-kubeconfig
KubeconfigMode: 0600
KubeCAFile:
SkipTLSVerify: false
K8sServiceProtocol:
K8sServiceHost: 10.233.0.1
K8sServicePort: 443
K8sNodeName: istio-cni-node-q7d24
UpdateCNIBinaries: true
SkipCNIBinaries: []
MonitoringPort: 15014
LogUDSAddress: /var/run/istio-cni/log.sock

2022-03-09T08:24:36.694166Z     info    CNI race repair configuration:
Enabled: true
RunAsDaemon: true
NodeName: sqkubworkelst06
LabelKey: cni.istio.io/uninitialized
LabelValue: true
DeletePods: true
LabelPods: true
SidecarAnnotation: sidecar.istio.io/status
InitContainerName: istio-validation
InitTerminationMsg:
InitExitCode: 126
LabelSelectors:
FieldSelectors:

2022-03-09T08:24:36.694517Z     info    Start a UDS server for CNI plugin logs
2022-03-09T08:24:36.694716Z     info    repair  Start CNI race condition repair.
2022-03-09T08:24:36.904931Z     info    install Copied istio-cni to /host/opt/cni/bin.
2022-03-09T08:24:36.905010Z     info    install Directory /host/secondary-bin-dir is not writable, skipping.
2022-03-09T08:24:36.905354Z     info    install write kubeconfig file /host/etc/cni/net.d/ZZZ-istio-cni-kubeconfig with:
# Kubeconfig file for Istio CNI plugin.
apiVersion: v1
kind: Config
clusters:
- name: local
  cluster:
    server: https://[10.233.0.1]:443
    certificate-authority-data: <CA cert from /var/run/secrets/kubernetes.io/serviceaccount/ca.crt>
users:
- name: istio-cni
  user:
    token: "<redacted>"
contexts:
- name: istio-cni-context
  context:
    cluster: local
    user: istio-cni
current-context: istio-cni-context

2022-03-09T08:24:36.905618Z     info    install Using CNI config template from CNI_NETWORK_CONFIG environment variable.
2022-03-09T08:24:36.905650Z     info    install CNI config: {
  "cniVersion": "0.3.1",
  "name": "istio-cni",
  "type": "istio-cni",
  "log_level": "info",
  "log_uds_address": "/var/run/istio-cni/log.sock",
  "kubernetes": {
      "kubeconfig": "/etc/cni/net.d/ZZZ-istio-cni-kubeconfig",
      "cni_bin_dir": "/opt/cni/bin",
      "exclude_namespaces": [ "istio-system", "kube-system" ]
  }
}
2022-03-09T08:24:36.905696Z     error   Installer exits with too many open files
2022-03-09T08:24:36.905704Z     info    install Cleaning up.
2022-03-09T08:24:36.905719Z     info    install Removing Istio CNI kubeconfig file: /host/etc/cni/net.d/ZZZ-istio-cni-kubeconfig
2022-03-09T08:24:36.905796Z     info    install Removing binary: /host/opt/cni/bin/istio-cni
Error: too many open files
```

On the node:

```shell
root@sqkubworkelst06:~# export LANG=C
root@sqkubworkelst06:~# ls -l /opt/cni/bin/istio-cni
ls: cannot access '/opt/cni/bin/istio-cni': No such file or directory
root@sqkubworkelst06:~# cat /proc/sys/fs/{file-nr,file-max,nr_open}
6432    0       9223372036854775807
9223372036854775807
1048576
root@sqkubworkelst06:~# head /proc/sys/fs/inotify/*
==> /proc/sys/fs/inotify/max_queued_events <==
16384

==> /proc/sys/fs/inotify/max_user_instances <==
128

==> /proc/sys/fs/inotify/max_user_watches <==
8192
```
It looks like the `too many open files` error is because inotify.

**Comment by sathieu** [comment]
I ended up by bumping the inotify maximums:

- fs.inotify.max_user_instances: '1024'  from '128'
- fs.inotify.max_user_watches: '1048576'  from '8192'


Ref: https://gitlab.com/kubitus-project/kubitus-installer/-/merge_requests/753
